# Exploring Hackers News Post

In this project, we're going to explore the status and two types of posts, Ask HN and Show HN, in Hackers News. 

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit.

Users submit "Ask HN" posts to ask the Hacker News community a specific question. Likewise, users submit "Show HN" posts to show the Hacker News community a project, product, or just generally something interesting.

To maximize the amount of comments, we want to determine:
1. which type of post receive more comments on average?
2. at what certain time do posts receive more comments on average?

It should be noted that our dataset has excluded all posts without comments.

## Understanding Data

In [1]:
#Import hacker_news.csv
import pandas as pd
hn = pd.read_csv('hacker_news.csv')

hn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20100 entries, 0 to 20099
Data columns (total 7 columns):
id              20100 non-null int64
title           20100 non-null object
url             17660 non-null object
num_points      20100 non-null int64
num_comments    20100 non-null int64
author          20100 non-null object
created_at      20100 non-null object
dtypes: int64(3), object(4)
memory usage: 1.1+ MB


In [2]:
hn.head()

id                                              title  \
0  12224879                          Interactive Dynamic Video   
1  10975351  How to Use Open Source and Shut the Fuck Up at...   
2  11964716  Florida DJs May Face Felony for April Fools' W...   
3  11919867       Technology ventures: From Idea to Enterprise   
4  10301696  Note by Note: The Making of Steinway L1037 (2007)   

                                                 url  num_points  \
0            http://www.interactivedynamicvideo.com/         386   
1  http://hueniverse.com/2016/01/26/how-to-use-op...          39   
2  http://www.thewire.com/entertainment/2013/04/f...           2   
3  https://www.amazon.com/Technology-Ventures-Ent...           3   
4  http://www.nytimes.com/2007/11/07/movies/07ste...           8   

   num_comments      author       created_at  
0            52    ne0phyte   8/4/2016 11:52  
1            10      josep2  1/26/2016 19:30  
2             1    vezycash  6/23/2016 22:20  
3             1     hswarna   6/17/2016 0:01  
4             2  walterbell   9/30/2015 4:12

## Calculating the Average Number of Comments for Ask HN and Show HN Posts

In [3]:
#extract posts that begin with either Ask HN or Show HN
hn['title'] = hn['title'].str.lower()
ask_hn = hn[hn['title'].apply(lambda x: x.startswith('ask hn'))]
show_hn = hn[hn['title'].apply(lambda x: x.startswith('show hn'))]

In [4]:
#count the mean of "ask hn" and "show hn" comments
avg_ask_comment = ask_hn['num_comments'].mean()
avg_show_comment = show_hn['num_comments'].mean()

print('avg_ask_comment: ', avg_ask_comment)
print('avg_show_comment: ', avg_show_comment)

avg_ask_comment:  14.038417431192661
avg_show_comment:  10.31669535283993


We can see that posts starting with "Ask hn" receive more comments that that of "Show hn" on average. 

## Finding the Amount of Ask Posts and Comments by Hour Created

Now we're going to determine if we can maximize the amount of comments an ask post receives by creating it at a certain hour. 

In [5]:
# create a table that shows average comments in every hour
import datetime as dt
ask_cmt_by_hr = ask_hn.loc[:, ['num_comments','created_at']]

# strift "hour" from the "create_at" column
ask_cmt_by_hr['created_at'] = pd.to_datetime(ask_cmt_by_hr['created_at'], format="%m/%d/%Y %H:%M")
ask_cmt_by_hr['created_at'] = ask_cmt_by_hr['created_at'].apply(lambda x: x.strftime('%H'))

ask_cmt_by_hr = ask_cmt_by_hr.groupby(['created_at']).mean()
ask_cmt_by_hr.sort_values(['num_comments'], ascending=False)

num_comments
created_at              
15             38.594828
02             23.810345
20             21.525000
16             16.796296
21             16.009174
13             14.741176
10             13.440678
14             13.233645
18             13.201835
17             11.460000
01             11.383333
11             11.051724
19             10.800000
08             10.250000
05             10.086957
12              9.410959
06              9.022727
00              8.127273
23              7.985294
07              7.852941
03              7.796296
04              7.170213
22              6.746479
09              5.577778

The top 5 hours for 'Ask HN' comments are:
02:00-03:00,
15:00-16:00,
16:00-17:00,
20:00-21:00,
21:00-22:00

The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

## Finding the Amount of Show Posts and Comments by Hour Created

In [6]:
# create a table that shows average comments in every hour
import datetime as dt
show_cmt_by_hr = show_hn.loc[:, ['num_comments','created_at']]

# strift "hour" from the "create_at" column
show_cmt_by_hr['created_at'] = pd.to_datetime(show_cmt_by_hr['created_at'], format="%m/%d/%Y %H:%M")
show_cmt_by_hr['created_at'] = show_cmt_by_hr['created_at'].apply(lambda x: x.strftime('%H'))

show_cmt_by_hr = show_cmt_by_hr.groupby(['created_at']).mean()
show_cmt_by_hr.sort_values(['num_comments'], ascending=False)

num_comments
created_at              
18             15.770492
00             15.709677
14             13.441860
23             12.416667
22             12.391304
12             11.803279
16             11.655914
07             11.500000
11             11.159091
03             10.629630
20             10.200000
19              9.800000
17              9.795699
09              9.700000
13              9.555556
04              9.500000
06              8.875000
01              8.785714
10              8.250000
15              8.102564
21              5.787234
08              4.852941
02              4.233333
05              3.052632

The top 5 hours for 'Post HN' comments are:
00:00-01:00
14:00-15:00
18:00-19:00
22:00-23:00
23:00-00:00

## Conclusion
In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) to reach the most engagement.

And while show posts did not receive comments as much as the ask posts did, if we want to release show posts for maximun amount of comments, the best time slot would be between 18:00 and 19:00.